### PR TITLE
Cache actor activations

### PR DESCRIPTION
**Pull request by richardartoul**
This makes the system significantly more efficient by making it so actor activations can be cached without compromising correctness which ensures we can tolerate many high volume actors.

Old Model
```
server = ensureActivation()
routeRPC(server)
```

This models guarantees external consistency, but requires reading a few keys from the registry for each RPC which makes high volume actors impractical.

New Model (client)
```
readVersion = fdb.getReadVersion() <-- batch/cache these
server = cache.getActivation()
routeRPC(server, readVersion)
```

New Model (serverr)
```
backgroundLoop:
    latestHeartbeatVersionStamp, heartbeatTTL = fdb.heartbeat()
server RPC handler:
    if req.versionStamp > lastHeartbeatVersionStamp + heartbeatTLL:
        reject()
```

TODO: Should prove this all out with a TLA+ model.

Previously the FDB-based implementation would not scale beyond ~ 4000 function calls/s globally. The new implementation trivially scales to 280,000 RPC/s locally:

```
Inputs
    invokeEvery 500ns
    numActors 10
Results
    numInvokes 4246528
    invoke/s 283101.86666666664
    median latency 0
    p95 latency 2.9742334234767007 ms
    p99 latency 13.874292528934026 ms
    p99.9 latency 29.080339799119038 ms
    max latency 64.72121241695221 ms

```

### CODE REVIEW

**Comment by aratz-lasa** [comment]
Amazing redesign! 

What command did you run for the benchmark? I would like to replicate it locally for playing around with it.

**Review by aratz-lasa — COMMENTED** [review]
Looks really good 👍

**Comment by richardartoul** [comment]
@aratz-lasa 

```
/opt/homebrew/bin/go test -timeout 60s -run ^TestBenchmarkFoundationRegistryInvoke$ github.com/richardartoul/nola/virtual
```

File: virtual/environment.go [hunk 1]
```diff
@@ -86,23 +128,48 @@ func (r *environment) Invoke(
...
+		ok              = false
+	)
+	if !r.opts.DisableActivationCache {
+		referencesI, ok = r.activationCache.Get(cacheKey)
+	}
+	if ok {
+		references = referencesI.([]types.ActorReference)
+	} else {
+		var err error
+		// TODO: Need a concurrency limiter on this thing.
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply virtual/environment.go:160]: yeah, will add one later

**Pull request merged**